# Task 4.8 - Grouping and Aggregating Data

### 1. Importing libraries and Data

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Define path

path = r'/Users/andymiller/Desktop/CareerFoundry Data Analytics/4. Python/Aug 2025 Instacart Basket Analysis'

In [3]:
# Import ords_prods_merge.pkl

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined_4.7.pkl'))

In [4]:
ords_prods_merge.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0             0             0   2539329        1             1   
1             0             0   2539329        1             1   
2             0             0   2539329        1             1   
3             0             0   2539329        1             1   
4             0             0   2539329        1             1   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   2                  8                     0.0         196   
1                   2                  8                     0.0       14084   
2                   2                  8                     0.0       12427   
3                   2                  8                     0.0       26088   
4                   2                  8                     0.0       26405   

   add_to_cart_order  ...  _merge Unnamed: 0_y  \
0                  1  ...    both          195   
1                  2  ...    both        14084   
2                  3  ...    both        12427   
3                  4  ...    both        26089   
4                  5  ...    both        26406   

                              product_name aisle_id  department_id  prices  \
0                                     Soda       77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk       91             16    12.5   
2                      Original Beef Jerky       23             19     4.4   
3               Aged White Cheddar Popcorn       23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls       54             17     1.0   

   merge_status    price_range_loc     busiest_day busiest_period_of_day  
0          both  Mid-range product  Regularly busy        Average orders  
1          both  Mid-range product  Regularly busy        Average orders  
2          both  Low-range product  Regularly busy        Average orders  
3          both  Low-range product  Regularly busy        Average orders  
4          both  Low-range product  Regularly busy        Average orders  

[5 rows x 21 columns]

In [5]:
#  #  There appears to be several redundant columns here that aren't in the examples in the exercise
#  Removing 'Unnamed: 0.1', 'Unnamed: 0_x', '_merge', 'Unnamed: 0_y'.

ords_prods_merge_2 = ords_prods_merge.drop(columns = (['Unnamed: 0.1', 'Unnamed: 0_x','_merge','Unnamed: 0_y']))


In [8]:
# Check that columns are gone

ords_prods_merge_2.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                     0.0       14084                  2          0   
2                     0.0       12427                  3          0   
3                     0.0       26088                  4          0   
4                     0.0       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  merge_status    price_range_loc     busiest_day busiest_period_of_day  
0         both  Mid-range product  Regularly busy        Average orders  
1         both  Mid-range product  Regularly busy        Average orders  
2         both  Low-range product  Regularly busy        Average orders  
3         both  Low-range product  Regularly busy        Average orders  
4         both  Low-range product  Regularly busy        Average orders

### 2. Find aggregated mean of 'order_number' grouped by 'department_id'

In [9]:
#  Aggregate mean order_number grouped by department_id

ords_prods_merge_2.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### 3. Comparing result to those of the subset taken previously

##### The results after analysing the entire dataframe are not that different to those when analysing only the first 1'000'000 rows.  The 1'000'000 rows was a fairly representative sample of the full data set.

### 4. Creating loyalty flags for existing customers

In [10]:
# Group by 'user_id' and add a column containing the max number of orders.

ords_prods_merge_2['max_order'] = ords_prods_merge_2.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/w2/ktpt6z_d1353vv2y_fgqj5cm0000gn/T/ipykernel_65466/1810770497.py:3: FutureWarning: The provided callable <function max at 0x106965080> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge_2['max_order'] = ords_prods_merge_2.groupby(['user_id'])['order_number'].transform(np.max)


In [18]:
ords_prods_merge_2.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      0.0         196                  1          0   
1                      0.0       14084                  2          0   
2                      0.0       12427                  3          0   
3                      0.0       26088                  4          0   
4                      0.0       26405                  5          0   
5                     15.0         196                  1          1   
6                     15.0       10258                  2          0   
7                     15.0       12427                  3          1   
8                     15.0       13176                  4          0   
9                     15.0       26088                  5          1   
10                    15.0       13032                  6          0   
11                    21.0         196                  1          1   
12                    21.0       12427                  2          1   
13                    21.0       10258                  3          1   
14                    21.0       25133                  4          0   

                               product_name  aisle_id  department_id  prices  \
0                                      Soda        77              7     9.0   
1   Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                       Original Beef Jerky        23             19     4.4   
3                Aged White Cheddar Popcorn        23             19     4.7   
4          XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                      Soda        77              7     9.0   
6                                Pistachios       117             19     3.0   
7                       Original Beef Jerky        23             19     4.4   
8                    Bag of Organic Bananas        24              4    10.3   
9                Aged White Cheddar Popcorn        23             19     4.7   
10                    Cinnamon Toast Crunch       121             14     4.0   
11                                     Soda        77              7     9.0   
12                      Original Beef Jerky        23             19     4.4   
13                               Pistachios       117             19     3.0   
14                    Organic String Cheese        21             16     8.6   

   merge_status    price_range_loc      busiest_day busiest_period_of_day  \
0          both  Mid-range product   Regularly busy        Average orders   
1          both  Mid-range product   Regularly busy        Average orders   
2          both  Low-range product   Regularly busy        Average orders   
3          bo

In [14]:
# Add the loyalty_flag column and classify each customer as Loyal, Regular or New

ords_prods_merge_2.loc[ords_prods_merge_2['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge_2.loc[(ords_prods_merge_2['max_order'] <= 40) & (ords_prods_merge_2['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge_2.loc[ords_prods_merge_2['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [19]:
ords_prods_merge_2.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      0.0         196                  1          0   
1                      0.0       14084                  2          0   
2                      0.0       12427                  3          0   
3                      0.0       26088                  4          0   
4                      0.0       26405                  5          0   
5                     15.0         196                  1          1   
6                     15.0       10258                  2          0   
7                     15.0       12427                  3          1   
8                     15.0       13176                  4          0   
9                     15.0       26088                  5          1   
10                    15.0       13032                  6          0   
11                    21.0         196                  1          1   
12                    21.0       12427                  2          1   
13                    21.0       10258                  3          1   
14                    21.0       25133                  4          0   

                               product_name  aisle_id  department_id  prices  \
0                                      Soda        77              7     9.0   
1   Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                       Original Beef Jerky        23             19     4.4   
3                Aged White Cheddar Popcorn        23             19     4.7   
4          XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                      Soda        77              7     9.0   
6                                Pistachios       117             19     3.0   
7                       Original Beef Jerky        23             19     4.4   
8                    Bag of Organic Bananas        24              4    10.3   
9                Aged White Cheddar Popcorn        23             19     4.7   
10                    Cinnamon Toast Crunch       121             14     4.0   
11                                     Soda        77              7     9.0   
12                      Original Beef Jerky        23             19     4.4   
13                               Pistachios       117             19     3.0   
14                    Organic String Cheese        21             16     8.6   

   merge_status    price_range_loc      busiest_day busiest_period_of_day  \
0          both  Mid-range product   Regularly busy        Average orders   
1          both  Mid-range product   Regularly busy        Average orders   
2          both  Low-range product   Regularly busy        Average orders   
3          bo

In [16]:
ords_prods_merge_2['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [17]:
# The loyalty flags have been successfully added

### 5. Check basic statistics for each loyalty category

In [24]:
# Look at the mean, median, min and max prices for items purchased by each loyalty group of customers

ords_prods_merge_2.groupby('loyalty_flag').agg({'prices': ['mean', 'median', 'min', 'max']})

prices                     
                       mean median  min      max
loyalty_flag                                    
Loyal customer    10.386336    7.4  1.0  99999.0
New customer      13.294670    7.4  1.0  99999.0
Regular customer  12.495717    7.4  1.0  99999.0

##### It looks as if new customers are attracted to the more expensive items averaging 13.29 per item whereas regular customers spend on average 0.79 less per item averaging 12.50.  Meanwhile the loyal customers average considerably less at just 10.39 per item.  The max price seems unrealistic and will require further investigation.

### 6. Create a spending flag for each user based on the average price across all their orders

In [26]:
#  First let's find the average item price for each user

ords_prods_merge_2.groupby('user_id').agg({'prices': ['mean', 'median', 'min', 'max']})

prices                  
             mean median  min   max
user_id                            
1        6.367797   4.70  1.0  14.0
2        7.515897   6.60  1.3  14.8
3        8.197727   8.20  1.3  14.4
4        8.205556   9.00  1.4  14.6
5        9.189189   8.60  3.2  14.8
...           ...    ...  ...   ...
206205   8.909375   9.05  1.4  23.8
206206   7.646667   6.70  1.1  14.8
206207   7.313453   6.80  1.2  24.7
206208   8.366617   8.00  1.1  21.6
206209   7.058915   5.90  1.1  18.8

[206209 rows x 4 columns]

In [11]:
#  Now that we can see some of these figures, we can add a column for the 'mean_item_price'

ords_prods_merge_2['mean_item_price'] = ords_prods_merge_2.groupby(['user_id'])['prices'].transform(np.mean)


/var/folders/w2/ktpt6z_d1353vv2y_fgqj5cm0000gn/T/ipykernel_65466/2976801014.py:3: FutureWarning: The provided callable <function mean at 0x106965a80> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge_2['mean_item_price'] = ords_prods_merge_2.groupby(['user_id'])['prices'].transform(np.mean)


In [33]:
ords_prods_merge_2.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                     0.0       14084                  2          0   
2                     0.0       12427                  3          0   
3                     0.0       26088                  4          0   
4                     0.0       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  merge_status    price_range_loc     busiest_day busiest_period_of_day  \
0         both  Mid-range product  Regularly busy        Average orders   
1         both  Mid-range product  Regularly busy        Average orders   
2         both  Low-range product  Regularly busy        Average orders   
3         both  Low-range product  Regularly busy        Average orders   
4         both  Low-range product  Regularly busy        Average orders   

   max_order  loyalty_flag  mean_item_price  
0         10  New customer         6.367797  
1         10  New customer         6.367797  
2         10  New customer         6.367797  
3         10  New customer         6.367797  
4         10  New customer         6.367797

In [12]:
# Add the spending_flag column and classify each customer as High spender or Low spender

ords_prods_merge_2.loc[ords_prods_merge_2['mean_item_price'] < 10, 'spender_flag'] = 'Low spender'

In [13]:
ords_prods_merge_2.loc[ords_prods_merge_2['mean_item_price'] >= 10, 'spender_flag'] = 'High spender'

In [14]:
ords_prods_merge_2.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                     0.0       14084                  2          0   
2                     0.0       12427                  3          0   
3                     0.0       26088                  4          0   
4                     0.0       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  merge_status    price_range_loc     busiest_day busiest_period_of_day  \
0         both  Mid-range product  Regularly busy        Average orders   
1         both  Mid-range product  Regularly busy        Average orders   
2         both  Low-range product  Regularly busy        Average orders   
3         both  Low-range product  Regularly busy        Average orders   
4         both  Low-range product  Regularly busy        Average orders   

   max_order  mean_item_price spender_flag  
0         10         6.367797  Low spender  
1         10         6.367797  Low spender  
2         10         6.367797  Low spender  
3         10         6.367797  Low spender  
4         10         6.367797  Low spender

In [15]:
ords_prods_merge_2['spender_flag'].value_counts(dropna = False)

spender_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

### 7. Creating a flag to distinguish frequent vs non-frequent customers

In [17]:
# Create a column median_days_since_prior_order

ords_prods_merge_2['median_days_since_prior_order'] = ords_prods_merge_2.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/w2/ktpt6z_d1353vv2y_fgqj5cm0000gn/T/ipykernel_65466/24842695.py:3: FutureWarning: The provided callable <function median at 0x106ab4fe0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge_2['median_days_since_prior_order'] = ords_prods_merge_2.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [38]:
ords_prods_merge_2.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                     0.0       14084                  2          0   
2                     0.0       12427                  3          0   
3                     0.0       26088                  4          0   
4                     0.0       26405                  5          0   

                              product_name  ...  prices  merge_status  \
0                                     Soda  ...     9.0          both   
1  Organic Unsweetened Vanilla Almond Milk  ...    12.5          both   
2                      Original Beef Jerky  ...     4.4          both   
3               Aged White Cheddar Popcorn  ...     4.7          both   
4         XL Pick-A-Size Paper Towel Rolls  ...     1.0          both   

     price_range_loc     busiest_day busiest_period_of_day max_order  \
0  Mid-range product  Regularly busy        Average orders        10   
1  Mid-range product  Regularly busy        Average orders        10   
2  Low-range product  Regularly busy        Average orders        10   
3  Low-range product  Regularly busy        Average orders        10   
4  Low-range product  Regularly busy        Average orders        10   

   loyalty_flag  mean_item_price spender_flag  median_days_since_prior_order  
0  New customer         6.367797  Low spender                           20.0  
1  New customer         6.367797  Low spender                           20.0  
2  New customer         6.367797  Low spender                           20.0  
3  New customer         6.367797  Low spender                           20.0  
4  New customer         6.367797  Low spender                           20.0  

[5 rows x 22 columns]

In [18]:
# Now add a frequent customer flag classiying each customer as "Non-frequent customer", 
# "Regular customer" or "Frequent customer".

ords_prods_merge_2.loc[ords_prods_merge_2['median_days_since_prior_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [19]:
ords_prods_merge_2.loc[(ords_prods_merge_2['median_days_since_prior_order'] > 10) & (ords_prods_merge_2['median_days_since_prior_order'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [20]:
ords_prods_merge_2.loc[ords_prods_merge_2['median_days_since_prior_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [21]:
# Checking Releavent Columns
ords_prods_merge_2[['user_id', 'mean_item_price', 'spender_flag', 'median_days_since_prior_order', 'order_frequency_flag']].head(15)

user_id  mean_item_price spender_flag  median_days_since_prior_order  \
0         1         6.367797  Low spender                           20.0   
1         1         6.367797  Low spender                           20.0   
2         1         6.367797  Low spender                           20.0   
3         1         6.367797  Low spender                           20.0   
4         1         6.367797  Low spender                           20.0   
5         1         6.367797  Low spender                           20.0   
6         1         6.367797  Low spender                           20.0   
7         1         6.367797  Low spender                           20.0   
8         1         6.367797  Low spender                           20.0   
9         1         6.367797  Low spender                           20.0   
10        1         6.367797  Low spender                           20.0   
11        1         6.367797  Low spender                           20.0   
12        1         6.367797  Low spender                           20.0   
13        1         6.367797  Low spender                           20.0   
14        1         6.367797  Low spender                           20.0   

   order_frequency_flag  
0      Regular customer  
1      Regular customer  
2      Regular customer  
3      Regular customer  
4      Regular customer  
5      Regular customer  
6      Regular customer  
7      Regular customer  
8      Regular customer  
9      Regular customer  
10     Regular customer  
11     Regular customer  
12     Regular customer  
13     Regular customer  
14     Regular customer

In [51]:
ords_prods_merge_2.shape

(32404859, 23)

In [23]:
# Check frequencies of flags

ords_prods_merge_2['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: count, dtype: int64

### 9. Export dataframe and pickle file

In [52]:
ords_prods_merge_2.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined_4.8.pkl'))